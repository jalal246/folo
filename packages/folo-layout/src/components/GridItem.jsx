import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import withcontext from "@folo/withcontext";

import { GridConsumer } from "./context";

import {
  CENTER,
  SPACE_BETWEEN,
  ROW,
  COLUMN,
  DISPLAY_FLEX,
  STRETCH
} from "../constants";

function location(from, to) {
  if (from !== null && to !== null) {
    return `${from} / ${to}`;
  }
  return `${from}`;
}

const propTypes = {
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

  row: PropTypes.number,
  toRow: PropTypes.number,

  col: PropTypes.number,
  toCol: PropTypes.number,
  isCenter: PropTypes.bool,

  style: PropTypes.objectOf(PropTypes.string),

  autoPositionCell: PropTypes.func.isRequired,

  isHorizontal: PropTypes.bool,
  children: PropTypes.node
};

const defaultProps = {
  component: "div",

  row: null,
  toRow: null,

  col: 0,
  toCol: null,
  isCenter: false,

  style: {},

  isHorizontal: true,
  children: null
};

/**
 * Used only when the Cell type is list
 * render option as defautl
 */
class GridItem extends PureComponent {
  state = {
    key: new Date().getTime()
  };

  render() {
    const {
      component: CellComponent,

      row,
      toRow,

      col,
      toCol,

      isCenter,
      isHorizontal,

      style: {
        display = DISPLAY_FLEX,

        flexDirection: fDirection,
        alignItems: aItems,
        ...otherStyle
      },

      autoPositionCell,

      children
    } = this.props;

    const { key } = this.state;

    // console.log('GridItem updated');
    const autoPosition = autoPositionCell({ key, row, toRow });

    const container = {
      display,

      ...(isHorizontal
        ? {
            flexDirection: fDirection || ROW,
            alignItems: aItems || CENTER
          }
        : {
            flexDirection: fDirection || COLUMN,
            alignItems: aItems || STRETCH
          }),

      justifyContent: isCenter ? CENTER : SPACE_BETWEEN,

      gridRow: location(autoPosition, toRow),
      gridColumn: location(col, toCol),

      ...otherStyle
    };

    return <CellComponent style={container}>{children}</CellComponent>;
  }
}

GridItem.propTypes = propTypes;
GridItem.defaultProps = defaultProps;

export { GridItem as PureGridItem };

export default withcontext({
  Component: GridItem,
  Consumer: GridConsumer,
  contextProps: ["autoPositionCell"]
});
