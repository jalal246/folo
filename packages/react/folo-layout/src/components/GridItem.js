import React from "react";
import positionStore from "../positionStore";

import {
  CENTER,
  SPACE_BETWEEN,
  ROW,
  COLUMN,
  DISPLAY_FLEX,
  STRETCH,
} from "../constants";

function location(from, to) {
  if (from !== null && to !== null) {
    return `${from} / ${to}`;
  }
  return `${from}`;
}

/**
 * For implicit grid
 * Takes column and row number
 * to inform the grid parent of total positions
 *
 * It collects numbers report it to Grid
 */
const GridItem = (props) => {
  const {
    component: CellComponent = "div",

    row = null,
    toRow = null,

    col = 0,
    toCol = null,

    isCenter = false,
    isHorizontal = true,

    style: {
      display = DISPLAY_FLEX,

      flexDirection: fDirection,
      alignItems: aItems,
      ...otherStyle
    } = {},

    id,
    children,

    ...rest
  } = props;

  const calculatedPosition = positionStore.autoPosition({
    key: id || `${new Date().getTime()}`,
    row,
    toRow,
  });

  const container = {
    display,

    ...(isHorizontal
      ? {
          flexDirection: fDirection || ROW,
          alignItems: aItems || CENTER,
        }
      : {
          flexDirection: fDirection || COLUMN,
          alignItems: aItems || STRETCH,
        }),

    justifyContent: isCenter ? CENTER : SPACE_BETWEEN,

    gridRow: location(calculatedPosition, toRow),
    gridColumn: location(col, toCol),

    ...otherStyle,
  };

  return (
    <CellComponent style={container} {...rest}>
      {children}
    </CellComponent>
  );
};

export default GridItem;
