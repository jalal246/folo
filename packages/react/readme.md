# React Components

> Folo React components

Directory content:

// TODO
