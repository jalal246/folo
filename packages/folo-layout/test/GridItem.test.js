import React from "react";
import { render, screen, cleanup } from "@testing-library/react";

import GridItem from "../src/components/GridItem";

const TEST_ID = "testForGrid";

const CENTER = "center";
const ROW = "row";
const COLUMN = "column";
const DISPLAY_FLEX = "flex";
const SPACE_BETWEEN = "space-between";
const STRETCH = "stretch";

const MyGridItem = (props = {}) => (
  <GridItem data-testid={TEST_ID} {...props} />
);

describe("GridItem - Implicit Layout", () => {
  afterEach(() => {
    cleanup();
  });

  describe("Testing Style", () => {
    it("Returns default style", () => {
      render(<MyGridItem />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns flexDirection column when it is isHorizontal = false", () => {
      render(<MyGridItem isHorizontal={false} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: STRETCH,
        flexDirection: COLUMN,
        // gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Overrides styles", () => {
      render(
        <MyGridItem
          style={{
            flexDirection: COLUMN,
            alignItems: STRETCH,
          }}
        />
      );
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: STRETCH,
        flexDirection: COLUMN,
        // gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns center", () => {
      render(<MyGridItem isCenter />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0",
        gridColumn: "0",
        justifyContent: CENTER,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("Testing gridRow", () => {
    it("Returns gridRow with toRow and alignItems center", () => {
      render(<MyGridItem toRow={6} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0 / 6",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridRow with toRow =0", () => {
      render(<MyGridItem toRow={0} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0 / 0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns gridRow with toRow =0", () => {
      render(<MyGridItem toRow={0} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0 / 0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("Testing gridColumn", () => {
    it("Returns column position with when col is provided", () => {
      render(<MyGridItem col={10} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0",
        gridColumn: "10",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("Returns column position with justifyContent, when toCol is provided", () => {
      render(<MyGridItem toCol={10} />);
      const { baseElement, style } = screen.getByTestId(TEST_ID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        // gridRow: "0",
        gridColumn: "0 / 10",
        justifyContent: SPACE_BETWEEN,
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });
});
