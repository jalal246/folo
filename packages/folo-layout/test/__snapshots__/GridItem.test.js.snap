// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem - Implicit Layout Testing Style Overrides styles 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing Style Returns center 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing Style Returns default style 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing Style Returns flexDirection column when it is isHorizontal = false 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing gridColumn Returns column position with justifyContent, when toCol is provided 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing gridColumn Returns column position with when col is provided 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing gridRow Returns gridRow with toRow =0 1`] = `undefined`;

exports[`GridItem - Implicit Layout Testing gridRow Returns gridRow with toRow =0 2`] = `undefined`;

exports[`GridItem - Implicit Layout Testing gridRow Returns gridRow with toRow and alignItems center 1`] = `undefined`;
